File: classwork/02-package-delivery-service/TwoDayPackage.cpp
#include "TwoDayPackage.h"

#include <iostream>

TwoDayPackage::TwoDayPackage()
    : Package{}, flatFee{} {}

TwoDayPackage::TwoDayPackage(double weight, double cost, const std::string& sName, const std::string& sZip, const std::string& rName, const std::string& rZip, double fee)
    : Package{weight, cost, sName, sZip, rName, rZip}, flatFee{fee} {}

double TwoDayPackage::getFlatFee() const {
    return flatFee;
}

void TwoDayPackage::setFlatFee(double fee) {
    flatFee = fee > 0.0 ? fee : 0.0;
}

double TwoDayPackage::calcCost() const {
    return Package::calcCost() + getFlatFee();
}

void TwoDayPackage::print() const {
    std::cout << "Two Day Package Details\n";
    printDetails();
    std::cout << "Flat Fee: $" << getFlatFee() << '\n';
    std::cout << "Calculated Cost: $" << calcCost() << '\n';
}

File: lab/cmps2232-lab-01-postal-system-demo/src/classes/package-two-day/TwoDayPackage.ts
// Two Day Package Class Implementation

import { Package } from '../package/Package';
import { ITwoDayPackage } from './ITwoDayPackage';

import { ShippingMethod } from '../../enums/shippingMethod';
import { PackageStatus } from '../../enums/packageStatus';

export class TwoDayPackage extends Package implements ITwoDayPackage {
  // constructor

  constructor(
    protected trackingNumber: number,
    protected senderName: string,
    protected senderAddress: string,
    protected receiverName: string,
    protected receiverAddress: string,
    protected weight: number,
    protected costPerUnitWeight: number,
    protected flatFee: number, // extra flat fee data member is declared here and set in constructor
    protected shippingMethod: ShippingMethod = ShippingMethod.TwoDay,
    protected status: PackageStatus = PackageStatus.Created
  ) {
    super(
      trackingNumber,
      senderName,
      senderAddress,
      receiverName,
      receiverAddress,
      weight,
      costPerUnitWeight,
      shippingMethod,
      status
    );
  }

  // extra methods

  getFlatFee(): number {
    return this.flatFee;
  }

  setFlatFee(flatFee: number): void {
    this.flatFee = flatFee;
  }

  // overridden methods

  calculateCost(): number {
    return (super.calculateCost() + this.flatFee) * 1.1;
  }

  printLabel(): void {
    console.log(`[Two Day Package Details]`);
    console.log(`Shipping Method: ${this.shippingMethod}`);
    console.log(`Status: ${this.status}`);
    console.log(
      `From ${this.senderName} (${this.senderAddress}) to ${this.receiverName} (${this.senderAddress})`
    );
    console.log(`Weight (lbs): ${this.weight}`);
    console.log(`Cost Per Unit Weight: $${this.costPerUnitWeight}`);
    console.log(`Flat Fee: $${this.flatFee}`);
    console.log(`Two Day Package Extra % Fee: 10%`);

    console.log(`Total Cost: $${this.calculateCost()}`);
  }
}
